### DISCUSSION TITLE
Welcome to astronomy Discussions!

### DISCUSSION THREAD

**Discussion by cosinekitty** [cosinekitty/astronomy#77] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by gerben-b** [comment]
Hi Don,

I seem to be the first here. My name is Gerben Bijhold and i am interested in moon and planet photography. For that purpose I built a web based app to control my camera and my telescope. The camera controls are fine now but the telescope control is still a challenge. In my app I also show the moon as it appears in the sky, but lately it has been too cloudy to check that. Hopefully, I will be able to do that soon.

Enough about me (or my app). I want you to now that I am quite impressed with your project here on github.

Cheers now,
Gerben

**Reply by cosinekitty** [reply]
Hi Gerben,

It is great to hear from you! Your camera and telescope control projects sound very interesting. I would like to learn more about them. I have always thought it would be fun to learn to photograph planets using the newer technology with CCD cameras and image correction software. Have you ever tried to discover new asteroids or comets?

Do you have any of your code published as open source?

Also, thank you for your kind words about Astronomy Engine. Please let me know if I can be of any help adapting it to any of the things you are working on. Also, I would be interested to know if there is anything missing from Astronomy Engine that would be useful to you.

Don

**Reply by gerben-b** [reply]
Hi Don,

I haven't published my code yet, but I intend to. I will let you know. I've never tried comets or asteroids but I may. Who knows. I am mainly interested in the moon and Jupiter's moons. Astrophotography in my location is hard since there is much light pollution where I live, unfortunately.

Regarding your question about anything missing, I am not sure if it is missing. In my web based app I also show the moon as it appears when I look out my window including it's tilt. I use suncalc.js for that and i didn't find the feature for the moons tilt in your code. Also I am not sure if I got it right since is has been cloudy ever since I created this.

My goal is to use my raspberry pi hq camera to find interesting spots on the moon and to help focus and then use the same camera for stills. That works pretty good now but I'm struggling to get my scope mount controlled from my web app. That is my next goal: guide my telescope to an object, precision guide to an interesting region, start tracking and then make pictures.

Cheers, Gerben

**Reply by cosinekitty** [reply]
That is an interesting idea for a feature: to calculate the tilt of the moon. Astronomy Engine already can calculate the moon's phase, and the horizontal coordinates of the moon for a given observer, but it does not report the tilt. In your opinion, what would be the the most helpful number? Perhaps it would be the clockwise angle of the sunlit side away from the local zenith? (I just added #90 for this; feel free to add comments there.)

I'm also interested in calculating Jupiter's moons (see #67). I have not yet found authoritative models and/or test data yet. I don't add anything to Astronomy Engine unless I have a way of proving the results are correct, using data from NASA/JPL, ESA, NOVAS, or some other authoritative source. So if you know of a good way of calculating Jupiter's moon and/or test data, I would love to learn that.

That's great that you are willing to publish your code when it is ready. I will look forward to studying it!

**Reply by gerben-b** [reply]
Hi Don,

My project is now published, it is called moon_cam. I haven't added readme's and attributions to your code yet but that will follow soon. It took me a while to understand how to publish. I'm not a software engineer, you know.

About your question what the most helpful number would be: I think a radian or an angle 0 - 360 should be okay as long as it is consistent with all your outputs en and as long as it is clear what the number means.

Let me know whet you think of moon_cam

Cheers, Gerben

**Reply by cosinekitty** [reply]
That's exciting to see the work I did used in a real project! It looks like you are using Astronomy Engine to calculate horizontal coordinates of Mars and Jupiter. Have you had any clear weather yet?

**Reply by gerben-b** [reply]
Unfortunately not, so I can't check, but there is a readme now in which i keep notes on such things. I designed it so that when a body's coordinates are displayed green, I should be able to see it out my window (south/west).
I haven't looked into data for Jupiters moons yet and i probably won't, but I know some astronomers that I've worked with and I'll ask them.

By the way, I am quite sure that your library is (should be) used in a fair number of projects since the quality is at a very professional level.

**Reply by gerben-b** [reply]
I've bees asking around among some astronomers whom I've worked with about the moons of Jupiter without results so far. They mentioned some tools though that may give you some inspiration:
- http://stellarium.org/
- https://www.astropy.org/
- http://simbad.u-strasbg.fr/simbad/
These are all well known tools, but as far as i can see not helpful for Jupiter's planets. Perhaps Stellarium's source code may help.

**Reply by cosinekitty** [reply]
OK, thanks for the information. I will take a look at Stellarium.

By the way, I am actively working on enhancements to Astronomy Engine to allow calculating the sunlit side of a body (the Moon or a planet) from the point of view of a camera aimed at that body. I hope to finish it this week. I realized in addition to the [orientation systems already supported](https://github.com/cosinekitty/astronomy/tree/master/source/js#coordinate-transforms), namely EQJ, EQD, ECL, HOR, I need to add one more that I will call "camera" or CAM. It will be based on horizontal coordinates (HOR) but with an arbitrary azimuth rotation angle, altitude lift angle, and lateral tilt angle. (I suspect in most cases people will want tilt = 0, but it costs little to make it completely general.)

Also, I am enhancing the [EquatorialCoordinates](https://github.com/cosinekitty/astronomy/tree/master/source/js#EquatorialCoordinates) class to include the cartesian vector, in addition to the existing spherical coordinates (RA, DEC, distance).

With both of these changes, I will then create a demo program that shows how to calculate the angle of the sunlit side of a body, as seen from a camera pointed at that body.

I will tag you on #90 when it is ready for you to try. Hopefully you can try it out and let me know how useful/correct it is.

**Reply by gerben-b** [reply]
I investigated a little more into the moons of Jupiter but found no useful pointers. This will be challenging i guess ;-)
When your new version with the CAM object is ready for testing, I will add it to my project and look out the window to the moon to verify its correctness.

**Reply by cosinekitty** [reply]
I am making progress on the camera idea. I have a `camera` branch in progress right now. I finished the C version, and I'm mostly done with the C# version. After that I will do JavaScript and Python.

**Reply by gerben-b** [reply]
Hi Don,

I was wondering if your engine can calculate RA/DEC with horizontal coordinates and a date/time as input. The reason why I ask is that I would like to be able to use altitude ans azimuth to control my telescope mount which only accepts RA/DEC parameters for it's goto function. With that i can use my equatorial mount for terrestrial applications.
If it's not possible, that's okay, then I'll just take another approach.

Thanks,
Gerben

**Reply by cosinekitty** [reply]
Yes, that is certainly possible. It sounds like you are aiming at objects on the Earth? Does your telescope computer correct for atmospheric refraction? That adds some complexity for objects near the horizon, which I'm guessing is very likely.

**Reply by gerben-b** [reply]
That sounds promising. I couldn't find it in your API description for astronomy.js and indeed, I will use it for targets on earth at close range using a camera with a standard lens. I won't need any correction for refraction so no worries there.

**Reply by cosinekitty** [reply]
OK, great. I will create a little demo program for you soon, because, although it is possible, it is certainly *not* obvious! I'm just finishing up the work on rotating the coordinate system for camera orientation.

**Reply by cosinekitty** [reply]
OK, I just merged all my changes into the master branch and published a new version of the npm package: `astronomy-engine@2.0.2`.

So now you have two new Node.js demo programs you can look at:

1. [equatorial.js](https://github.com/cosinekitty/astronomy/blob/master/demo/nodejs/equatorial.js) gives an example of converting horizontal coordinates to equatorial coordinates. It actually prints out 4 solutions, with/without refraction, and using J2000/of-date equatorial coordinates. You will need to experiment with your equipment to see which system is the best fit. The equatorial.js program includes self-checking, where I reverse the calculation using a different (and well-tested) algorithm, to prove it ends up exactly with the same (alt, az) you started with.
2. [camera.js](https://github.com/cosinekitty/astronomy/blob/master/demo/nodejs/camera.js) is a sample program that shows how to calculate the angle of the sunlit side of the Moon, as seen by a level camera facing the Moon's azimuth and tilted upward so the view is centered on the Moon's altitude. It also calculates some other things you might find interesting. This one's testing is rather weak, because I have no idea how to find authoritative data for it. But I did go outside and look at the Moon and at least it wasn't horribly wrong, LOL!

You will need to use version 2.0.2 of the npm package, or the latest code in the master branch, for these programs to work.

I hope this helps! Please let me know if you have any questions or run into any issues.

**Reply by gerben-b** [reply]
Hey Don,

You are fast :-) I just installed your latest version and ran the equatorial demo. Is the range of the output RA/DEC from -180 to +180?

Thanks,
Gerben

**Reply by cosinekitty** [reply]
RA is right ascension: 0 to 24 sidereal hours.  DEC is declination in degrees north/south, so it goes from -90 to +90.

**Reply by gerben-b** [reply]
Ofcourse, I could have known that it is in your API description. Thanks.

I tried the equatorial function and it seems that the browser version produces a different outcome than the node version. I'm not sure yet so I keep testing. I may do something wrong myself.

The camera object seems to produce the correct value. It is going to take a while to integrate this in moon_cam but I'll keep you informed.

**Reply by gerben-b** [reply]
I made a mistake. The outcome of both versions is the same and it appears to work perfectly. The coming days, I'll integrate things in moon_cam and then start really testing. My telescope mount already moves like a charm with your module and hopefully in the correct direction :-) Also I looked some more into Stellarium and indeed it knows where Jupiter's moons are. As soon as Jupiter becomes visible in my region I'm going to test that too.

**Reply by cosinekitty** [reply]
That's great news! I will be excited to find out how it's working for your tests. I will start looking at Stellarium's source code to figure out how to calculate Jupiter's four major moons. Maybe they even have references so I can learn where they got their formulas from.

**Reply by meshula** [reply]
Chapter 44 of Jean Meeus' Astronomical Algorithms, Second Ed., is "Satellites of Jupiter". Chapter 46 has Saturn's.

I highly recommend adding this book to your library if you haven't got it already.

**Reply by cosinekitty** [reply]
@meshula Thanks for the tip! I will see if I can acquire a copy of that book. (I have certainly heard of Meeus; in fact Astronomy Engine uses the Espenak/Meeus model for the rate at which the Earth's rotation speed changes over the centuries.)

**Reply by cosinekitty** [reply]
Update: I just check Amazon, and the cheapest I can find that book is over $300 US.  WOW!!! It's a collector's item...

**Reply by meshula** [reply]
OMG, I'm lucky to have it in my library :\
Code for Chapter 44 is here - it's easy to read :) https://github.com/soniakeys/meeus/tree/master/v3/jupitermoons

and here's chapter 46 - https://github.com/soniakeys/meeus/tree/master/v3/saturnmoons

**Reply by cosinekitty** [reply]
That is perfect!  I will see if I can build that Go code and generate some test data.

**Reply by cosinekitty** [reply]
@meshula I looked at the soniakeys/meeus repo. The code for Jupiter's moons looks promising. It has two different models, a simple one and a complicated one. Both calculate the 4 major moons relative to Jupiter's equatorial plane. The simple one treats the moons as always exactly in that equatorial plane, calculating x and y values only, with z=0 always. The complicated one includes their slight movement above and below the plane.

Because you have the Meeus book and I don't, I'm wondering how I can convert Jupiter's equatorial plane into another coordinate system that Astronomy Engine can handle. Is there some mention of how to convert the Jupiter frame of reference into Earth's equatorial plane, or the plane of the ecliptic? In other words, I could translate her code from Go into other languages, and get some numbers, but I don't know how they relate to any of the existing things I calculate. I hope I'm making sense...

**Reply by gerben-b** [reply]
> Update: I just check Amazon, and the cheapest I can find that book is over $300 US. WOW!!! It's a collector's item...

I went looking too, and found them cheaper, e.g.: https://www.willbell.com/math/mc1.HTM. I even found a free pdf, but I'm not sure if that is legal, so I ignored that.

Anyway, the proceedings seem promising. As soon as Jupiter is visible from my window I'll start observing those 4 moons to see if I can capture their movement with a time lapse of a couple of hours.

**Reply by cosinekitty** [reply]
I clicked on that link, but then I see: "Willmann-Bell is closed for business. Orders will no longer be accepted."

**Reply by meshula** [reply]
To make it easier to follow, I started a new thread for Jupiter's Moons with a couple of breadcrumbs, including a link to the specific calculations in Chapter 43 needed to compute the variables needed for Chapter 44, and a description of the coordinate system. https://github.com/cosinekitty/astronomy/discussions/97
